### ISSUE TITLE
add a sequential run order in with_columns execution

### ISSUE THREAD

**Issue by tikendraw** [pola-rs/polars#9062]
### Problem description

I am writing to express my admiration for the remarkable speed at which Polars executes operations using the with_columns method. It has significantly enhanced my data processing workflows. However, I have encountered scenarios where I require sequential execution of specific operations, instead of parallel execution. Currently, achieving this sequential behavior involves repeatedly using the `df.with_columns` method, which can be tedious and less intuitive.

Therefore, I would like to propose the addition of a new feature to Polars that enables sequential execution of operations within the with_columns method. This feature would provide greater flexibility and ease of use for users like myself, who require sequential computation.

I have two suggestions for implementing this feature:

**Option 1: Using a Run Sequence List**

- Introduce a new parameter, let's call it run_sequence, that accepts a list of indices representing the order in which the expressions should be executed.

    For example, if `run_sequence = [1, 3, 2]`, Polars would execute the expressions in the following order: expression 1, expression 3, and then expression 2.

or 

**Option 2: Using a Boolean Flag**

- Alternatively, introduce a boolean parameter, let's call it run_sequentially, which, when set to True, would enforce sequential execution of expressions based on their order in the with_columns method.

Both options would provide users with a more convenient and expressive way to control the execution order of operations within the with_columns method, avoiding the need for repetitive calls.

**Comment by zundertj** [comment]
It may sometimes be faster to simply re-use expressions, and thus duplicate computations, rather than executing in sequence. That would be assuming the operations can run together at the same speed as individually. I.e. suppose you have:

```python
# given dataframe with a column "a", you want this in sequence:
df.with_columns(a2 = pl.col("a")**2, a2_relative=pl.col("a2")/pl.col("a2").sum())

# instead we duplicate the a2 computation like this:
expr = pl.col("a")**2
df.with_columns(a2=expr, a2_relative=expr / expr.sum())
```
Having said that, there are cases where this would be slower, or where the logic is such that doing it in a single expression is impossible (applying `over` twice for instance), so I do understand the request.

I cannot judge if this is possible at all, but if it is, I would strongly suggest to keep the api simple and go for the boolean flag, option 2, and by default off.  Option 1 is a bit too much "magic numbers" for me.

**Comment by alexander-beedie** [comment]
Presumably the main (only?) reason to run operations in sequence is because you want one to depend on the results of a column that was just created in the same `with_columns` scope? 

Something like:
```python
df = pl.DataFrame( {'x':1} )

df.with_columns(
  y = pl.col("x") * 2,
  z = pl.col("y") * 3,  # requires that 'y' is created before it can be referenced
)
```
If such a mode was to be supported, it would probably be better-handled by generating a DAG (directed acyclic graph) of the column dependencies rather than a manual parameter - such dependencies could be pretty determined with a basic topological sort.

Alternatively, rather than go for something exotic, you could also just `pipe` your functions; a small wrapper could take care of the boilerplate involved in wrapping with a lambda, and then you could register it with the [extension](https://pola-rs.github.io/polars/py-polars/html/reference/api.html) API for continued use.

I've done something similar to this before; I might write a proof-of-concept DAG version for use with the extension API for a bit of fun later, but I'm going to be travelling for a few days from tomorrow, so won't get to it for a bit 🤔

**Comment by ritchie46** [comment]
I don't want to complexity in the API. In the future we can fix this on the backend by common-subexpression-elimination.

Currently you can cache manually or simply split the two with_columns calls.

There is an easy way around this. No need for us to add complexity on the API side.

**Comment by alexander-beedie** [comment]
> There is an easy way around this. No need for us to add complexity on the API side.

Indeed; I think it makes for a great extension rather than a core feature - and I couldn't help myself, so I've already got a working `dag` prototype up & running that determines the evaluation order automatically via a topological sort against the expression root names... ;)

```python 
df = pl.DataFrame({'x': 1})

df.dag.with_columns(
    q = pl.col("x") * pl.col("y"),
    z = pl.col("y") * 3,
    y = pl.col("x") * 2,
    m = pl.lit(0),
    n = pl.col('x') + 1
)
# ┌─────┬─────┬─────┬─────┬─────┬─────┐
# │ x   ┆ q   ┆ z   ┆ y   ┆ m   ┆ n   │
# │ --- ┆ --- ┆ --- ┆ --- ┆ --- ┆ --- │
# │ i64 ┆ i64 ┆ i64 ┆ i64 ┆ i32 ┆ i64 │
# ╞═════╪═════╪═════╪═════╪═════╪═════╡
# │ 1   ┆ 2   ┆ 6   ┆ 2   ┆ 0   ┆ 2   │
# └─────┴─────┴─────┴─────┴─────┴─────┘
```
I think this approach is the way to go, as non-trivial dependency chains shouldn't be managed manually.

I'll finish this off when I come back from travelling and put it up on GitHub somewhere; probably makes for a good LinkedIn article on our extension API too, heh.

(We did a lot of work with graph dependencies at my old shop, so I know there are plenty of interesting use-cases).

**Comment by mcrumiller** [comment]
I manually implemented this at work as well uisng a simple depth-first sort, so having an easily accessible extension would be great.

**Comment by ritchie46** [comment]
I am interested in the algorithm. Maybe I put it in the expression engine.

**Comment by alexander-beedie** [comment]
> I am interested in the algorithm. Maybe I put it in the expression engine.

Nice. I'll finish it up when I'm back from travelling this week and ping you - algorithmically it's little more than a vanilla non-recursive topological sort.

**Comment by avimallu** [comment]
I've not had the need to use recurrent sequential execution in with_columns expression more than once, but I would have very much needed to use it in the past. Avoiding multiple calls to `with_columns` would be great then, even if at the cost of parallelism.

**Comment by gab23r** [comment]
@alexander-beedie, do you have any news on the one ? I think this would be super valuable for polars. Moreover, I think it could be useful the other way around. Today if you do `with_columns(a = 1).with_columns(b = 1)` it is done sequentially but polars could understand that it can actually be done in parallel. What do you think ?

**Comment by mcrumiller** [comment]
I've thought a bit more about this one and I'm a little iffy on it. I think there are some ambiguous cases, namely overwriting an existing column. Super simple example:

```python
import polars as pl
from polars import col

df = pl.DataFrame({'a': [1, 2, 3]})

df.with_columns(
    (col('a') + 1).alias('a'),
    when(col('a') < 2).then(1).otherwise(2).alias('b')
)
```

If we created a dependency tree here w/ execution order, would column `b` depend on the updated `a` or or the old `a`? I would be inclined to say the latter. I guess this isn't too complicated, I was envisioning a complicated ambiguous case but I'm now unable to come up with one.

**Comment by gab23r** [comment]
You are right, this will be a breaking change. But I think the only way to do it is to say that b depend on the new a. The rule would be: "everything works as if things were done sequentially" but then the optimizer do its job to parallelize what he can.

**Comment by ritchie46** [comment]
I don't want this extra complexity in polars. If you want to use an update column, you need two `with_columns`. This makes it much more readable, simple, and explainable.

**Comment by tikendraw** [comment]
Hello, Do you guys think this is a good feature ?

**Comment by alexander-beedie** [comment]
> Hello, Do you guys think this is a good feature ?

Not in the core code, no. I have actually written the code to do this as an extension instead, but haven’t pushed it out anywhere yet. I’ll try and finish that up this weekend and put a link to it here.

**Comment by ion-elgreco** [comment]
> I've thought a bit more about this one and I'm a little iffy on it. I think there are some ambiguous cases, namely overwriting an existing column. Super simple example:
> 
> ```python
> import polars as pl
> from polars import col
> 
> df = pl.DataFrame({'a': [1, 2, 3]})
> 
> df.with_columns(
>     (col('a') + 1).alias('a'),
>     when(col('a') < 2).then(1).otherwise(2).alias('b')
> )
> ```
> 
> If we created a dependency tree here w/ execution order, would column `b` depend on the updated `a` or or the old `a`? I would be inclined to say the latter. I guess this isn't too complicated, I was envisioning a complicated ambiguous case but I'm now unable to come up with one.

Could disambiguate it by using other expressions for within context only columns and global columns.

**Comment by magarick** [comment]
This thread exposes what I see as a core tension in building data tools. You have two main groups of people who use them. For simplicity, let's call them scientists and engineers. I'm generalizing a bit, but I think it's _mostly_ true. For "scientists" the end result is the analysis or model, the work is interactive, and quick iteration and conciseness are valued. For "engineers" the end result is pipelines and transformed data, the work is in building and maintaining systems, and explicitness and robustness are valued.

So it makes sense to me that when you're doing one-off interactive analyses, you prefer a bit of "magic" and "just keep going" as opposed to the "be very explicit" and "fail early and loudly" that you need when automating. In each context things like readability and simplicity really do have different meanings. I think Polars, however, is unique in that it can be very good for both use cases.

Since, Polars expressions are so flexible and masterfully work around Python's foolish lack of real metaprogramming, it's pretty easy to build concise tools on top. In fact, I've fiddled around with something which might be a good home for this type of feature request https://github.com/magarick/tarsiers. I'm saying Python because I can't imagine this type of feature being particularly useful on the Rust side.

@ritchie46 I understand your desire to keep the core Polars API simple and explicit. So what do you think about having a semi-official "terse" API that's not part of the main library and is totally opt-in? I do think having as much as possible in one place is better than potentially a load of dispersed one-off extensions of varying quality and maintenance.

@alexander-beedie If you're looking for a place to put your extension, would you be interested in Tarsiers?

**Comment by ion-elgreco** [comment]
> This thread exposes what I see as a core tension in building data tools. You have two main groups of people who use them. For simplicity, let's call them scientists and engineers. I'm generalizing a bit, but I think it's _mostly_ true. For "scientists" the end result is the analysis or model, the work is interactive, and quick iteration and conciseness are valued. For "engineers" the end result is pipelines and transformed data, the work is in building and maintaining systems, and explicitness and robustness are valued.
> 
> So it makes sense to me that when you're doing one-off interactive analyses, you prefer a bit of "magic" and "just keep going" as opposed to the "be very explicit" and "fail early and loudly" that you need when automating. In each context things like readability and simplicity really do have different meanings. I think Polars, however, is unique in that it can be very good for both use cases.
> 
> Since, Polars expressions are so flexible and masterfully work around Python's foolish lack of real metaprogramming, it's pretty easy to build concise tools on top. In fact, I've fiddled around with something which might be a good home for this type of feature request https://github.com/magarick/tarsiers. I'm saying Python because I can't imagine this type of feature being particularly useful on the Rust side.
> 
> @ritchie46 I understand your desire to keep the core Polars API simple and explicit. So what do you think about having a semi-official "terse" API that's not part of the main library and is totally opt-in? I do think having as much as possible in one place is better than potentially a load of dispersed one-off extensions of varying quality and maintenance.
> 
> @alexander-beedie If you're looking for a place to put your extension, would you be interested in Tarsiers?

As a 'scientist' who also does engineering work.  I can only see such semi-official APIs create more work in the end for engineers. You will have to rewrite most code back to the official API to keep it consistent.

**Comment by magarick** [comment]
> As a 'scientist' who also does engineering work. I can only see such semi-official APIs create more work in the end for engineers. You will have to rewrite most code back to the official API to keep it consistent.

I couldn't disagree with this more. The vast majority of code written in this way would never see a production system. Further, going from "research mode" to "production" mode always entails a lot more than copy-pasting code out of a notebook. Everything is written on top of the existing API so translation is easy, and it's not like I'm proposing something that will yield a gigantic mess that couldn't be usable in production. Whether this introduced new work at all would come down more to organizational policies than anything else.
Also, it seems like the implicit assumption is that you should force a large number, maybe even a majority of your users to accept an interface that's not appropriate for their tasks.

**Comment by avimallu** [comment]
> > As a 'scientist' who also does engineering work. I can only see such semi-official APIs create more work in the end for engineers. You will have to rewrite most code back to the official API to keep it consistent.
> 
> I couldn't disagree with this more. The vast majority of code written in this way would never see a production system. Further, going from "research mode" to "production" mode always entails a lot more than copy-pasting code out of a notebook. Everything is written on top of the existing API so translation is easy, and it's not like I'm proposing something that will yield a gigantic mess that couldn't be usable in production. Whether this introduced new work at all would come down more to organizational policies than anything else.
> Also, it seems like the implicit assumption is that you should force a large number, maybe even a majority of your users to accept an interface that's not appropriate for their tasks.

Perhaps a balance that can avoid the "dplyr for exploration, data.table for production", or the equivalent "pandas for exploration, spark for production" paradigm is to have these tools have a feature that "translates" the "scientist" API to native Polars API and have the _fail loudly_ part for those sections that don't have official API support?

**Comment by ritchie46** [comment]
I will close this as I want the semantics of a context to remain as is.

**Comment by henryharbeck** [comment]
Hi @alexander-beedie, any chance of sharing whatever you've currently got for this?
Would be interested in trying it out as an extension, and also curious to see the implementation.
All good if not fully complete, happy to finish it myself.

> > Hello, Do you guys think this is a good feature ?
> 
> Not in the core engine, no. I have actually written the code to do this as an extension instead, but got sidetracked and haven’t pushed it out anywhere yet. I’ll try and finish that up this weekend (or shortly!) and put a link to it here.

nb: Not implying to re-open this issue

**Comment by deanm0000** [comment]
I thought hacking at this might be fun and came up with 

```python
def aug_exprs(*exprs, **named_exprs):
    no_aliases = {}
    for expr in exprs:
        name = expr.meta.output_name()
        no_aliases[name] = expr.meta.undo_aliases().meta.serialize(format="json")
    for name, expr in named_exprs.items():
        no_aliases[name] = expr.meta.undo_aliases().meta.serialize(format="json")
    no_sub = True
    while no_sub:
        for one in no_aliases.keys():
            for name, expr in no_aliases.items():
                if name == one:
                    continue
                str_col = "{" + f'"Column":"{one}"' + "}"
                if str_col in expr:
                    no_sub = False
                    expr = expr.replace(str_col, no_aliases[one])
                    no_aliases[name] = expr

    return [
        pl.Expr.deserialize(StringIO(expr), format="json").alias(name)
        for name, expr in no_aliases.items()
    ]
```

I'm sure there are a ton of places that this fails but essentially, it goes inside of a select or with_columns and your expressions go inside of this. Then this looks at all your expressions to see if the output names of any of them are used by any of the other ones. If that happens it just copy/pastes the expression from one in place of the reference to the other. It does that loop over and over until it can't make any substitutions. Then it just returns the result.

It looks like this

```python
df=pl.DataFrame({"a":[1,2,3]})

df.select(aug_exprs((pl.col("a")*2).alias("b"), (pl.col("b")+2).alias("c"), (pl.col("c")-3).alias("d")))
shape: (3, 3)
┌─────┬─────┬─────┐
│ b   ┆ c   ┆ d   │
│ --- ┆ --- ┆ --- │
│ i64 ┆ i64 ┆ i64 │
╞═════╪═════╪═════╡
│ 2   ┆ 4   ┆ 1   │
│ 4   ┆ 6   ┆ 3   │
│ 6   ┆ 8   ┆ 5   │
└─────┴─────┴─────┘
```
